rotkehlchen/chain/decoding/decoder.py

```diff
@@ -0,0 +1,365 @@
+
+        refresh_balances, new_events = False, []
+        total_transactions = len(tx_hashes)
+        log.debug(f'Started logic to decode {total_transactions} transactions from {self.chain_name}')  # noqa: E501
+        for tx_index, tx_hash in enumerate(tx_hashes):
+            log.debug(f'Decoding logic started for {tx_hash} ({self.chain_name})')
+            if send_ws_notifications and tx_index % 10 == 0:
+                log.debug(f'Processed {tx_index} out of {total_transactions} transactions from {self.chain_name}')  # noqa: E501
+                self.msg_aggregator.add_message(
+                    message_type=WSMessageType.PROGRESS_UPDATES,
+                    data={
+                        'chain': self.chain_name,
+                        'subtype': str(ProgressUpdateSubType.UNDECODED_TRANSACTIONS),
+                        'total': total_transactions,
+                        'processed': tx_index,
+                    },
+                )
+
+            # TODO: Change this if transaction filter query can accept multiple hashes
+            with self.database.conn.read_ctx() as cursor:
+                context = self._load_transaction_context(
+                    cursor=cursor,
+                    tx_hash=tx_hash,
+                )
+
+            fresh_events, new_refresh_balances, reload_decoders = self._decode_transaction_from_context(  # noqa: E501
+                context=context,
+                ignore_cache=ignore_cache,
+                delete_customized=delete_customized,
+            )
+
+            if events is not None:
+                events.extend(fresh_events)
+
+            new_events.extend(fresh_events)
+            if new_refresh_balances is True:
+                refresh_balances = True
+
+            if reload_decoders is not None:
+                with self.database.conn.read_ctx() as cursor:
+                    self.reload_specific_decoders(cursor, decoders=reload_decoders)
+
+        if send_ws_notifications:
+            self.msg_aggregator.add_message(
+                message_type=WSMessageType.PROGRESS_UPDATES,
+                data={
+                    'chain': self.chain_name,
+                    'subtype': str(ProgressUpdateSubType.UNDECODED_TRANSACTIONS),
+                    'total': total_transactions,
+                    'processed': total_transactions,
+                },
+            )
+
+        return refresh_balances, new_events
+
+    @abstractmethod
+    def _calculate_gas_burned(self, tx: T_Transaction) -> FVal:
+        """Calculates gas burn based on relevant chain's formula."""
+
+    def _handled_decode_exceptions(self) -> tuple[type[Exception], ...]:
+        """Return exceptions this decoder treats as handled
+        (subclasses extend for chain-specific errors)."""
+        return (
+            UnknownAsset,
+            WrongAssetType,
+            DeserializationError,
+            IndexError,
+            ValueError,
+            ConversionError,
+        )
```

Comment:
why have this as a function instead of a variable in the file or the decoder itself? It's an extra function call for every decode